message view: Show message source button by default.

Signior-X opened the pull request:
This commit removes the unless msg/locally_echoed condition for the
edit content div to be shown. Also adds the View source button by
default to the edit_content div. This ensures that the message source
can be seen if a very long message has been sent and it fails due to
certain reasons.

Fixes #17650

<!-- What's this PR for?  (Just a link to an issue is fine.) -->
[CZO Conversations here](https://chat.zulip.org/#narrow/stream/2-general/topic/View.20source.20for.20failed.20messages/near/1134083)

**Testing plan:** <!-- How have you tested? -->
Tested locally using the developer tools to fail the message sent.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
A screenshot of the source button coming for the failed message.
![image](https://user-images.githubusercontent.com/56730716/111402974-2e2f6e00-86f2-11eb-82cc-220c52fc00ee.png)

And it opens.
![image](https://user-images.githubusercontent.com/56730716/111403061-54eda480-86f2-11eb-982b-d75a8923e5d6.png)

Also when the message is successfully sent, the source changes to the edit button as what the default behavior is.


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

YashRE42 commented:
This works, although I'm not sure if we want to add to the `edit_content` div.

The patch bellow also seems to work:
```diff
diff --git a/static/js/ui_init.js b/static/js/ui_init.js
index ee0e415904..97a8850b6c 100644
--- a/static/js/ui_init.js
+++ b/static/js/ui_init.js
@@ -94,10 +94,6 @@ function message_hover(message_row) {
     if (current_message_hover && rows.id(current_message_hover) === id) {
         return;
     }
-    // Don't allow on-hover editing for local-only messages
-    if (message_row.hasClass("local")) {
-        return;
-    }
     const message = current_msg_list.get(rows.id(message_row));
     message_unhover();
     current_message_hover = message_row;
diff --git a/static/templates/message_controls.hbs b/static/templates/message_controls.hbs
index 5b887263e9..2884752961 100644
--- a/static/templates/message_controls.hbs
+++ b/static/templates/message_controls.hbs
@@ -1,8 +1,6 @@
 <div class="message_controls{{#status_message}} sender-status-controls{{/status_message}} no-select">
     {{#if msg/sent_by_me}}
-        {{#unless msg/locally_echoed}}
-        <div class="edit_content"></div>
-        {{/unless}}
+    <div class="edit_content"></div>
     {{/if}}
 
     {{#unless msg/sent_by_me}}
````

I'd wait for review before making changes though

timabbott commented:
Yeah, I think roughly what @YashRE42 has in mind seems like a good direction; we just need to make sure that we're not going to create a secondary problem where we offer to let you edit content with a pencil while waiting for a message to send (which is a feature we should build but does not yet exist -- so would be a good follow-up project once we've finished this, perhaps in stages starting with ("let you start editing but not allow submission until the server's ACKed the message" and then something that would actually queue your edit to be sent to the server as soon as the message is ACKed by the server).

Signior-X commented:
Thanks @YashRE42 for the patch. Yes, this also works. This adds the message source using the javascript, so it is good as suggested by @timabbott that we can create a feature in the future regarding editing the message not sent also.

>  we're not going to create a secondary problem where we offer to let you edit content with a pencil while waiting for a message to send.

The edit isn't enabled. The `is_topic_editable` perfectly ensures this and only the markdown source is shown only.

timabbott commented on static/js/ui_init.js:
```diff
@@ -94,10 +94,7 @@ function message_hover(message_row) {
+
     const message = current_msg_list.get(rows.id(message_row));
     message_unhover();
     current_message_hover = message_row;
```
Maybe worth adding a comment next to the `is_topic_editable` check noting that locally echoed messages fit into that code path as well?

Signior-X replied:
Sure, this will be better to track down.

Signior-X commented:
@timabbott I have updated the PR adding a comment before the `is_topic_editable` check noting that locally echoed messages fit into that code path now also added the issue number so that it can be easily trackable when that change happened and will be easily trackable/understandable in future.

timabbott commented:
I tweaked the comment to be more clear and then merged this.

@Signior-X if you want a good next project, I think we should add a "Convert to draft" button next to the "resend / cancel" buttons that basically moves the message back into the compose box.  Not sure exactly what we want -- opened https://github.com/zulip/zulip/issues/17697 to track the issue.

timabbott commented:
Oh, one more thing -- does the vertical alignment and spacing of the buttons there seem a bit off?  I suspect that the red icons are a couple pixels off, probably because they were missed in some sweep where we aligned the other message actions icons more nicely with the timestamp.  Can you try doing a quick PR to fix that?

Signior-X commented:
> @Signior-X if you want a good next project, I think we should add a "Convert to draft" button next to the "resend / cancel" buttons.

That would be a really great idea and these days as I am focusing more on Compose area, this will going to be fun.

> Oh, one more thing -- does the vertical alignment and spacing of the buttons there seem a bit off? I suspect that the red icons are a couple pixels off, probably because they were missed in some sweep where we aligned the other message actions icons more nicely with the timestamp. Can you try doing a quick PR to fix that?

Yes sure, doing this right now.

Signior-X commented:
@timabbott created the PR fixing the alignment.
